alternator: allow parallel scan

psarna opened the pull request:
Parallel scans can be performed by providing Segment and TotalSegments
attributes to Scan request, which can be used to split the work among
many workers.
This test makes the parallel scan test succeed, so the xfail is removed.

Fixes #5059

nyh commented on alternator/executor.cc:
```diff
@@ -2951,11 +2951,38 @@ static future<executor::request_return_type> do_query(schema_ptr schema,
+static dht::token token_for_segment(int segment, int total_segments) {
+    assert(total_segments > 1 && segment < total_segments);
+    int64_t delta = std::numeric_limits<int64_t>::max() / total_segments * 2;
+    return dht::token::from_int64(std::numeric_limits<int64_t>::min() + 1 + delta * segment);
```
I slept under a rock, and suddenly tokens are *always* 64 bits?

nyh replied:
And where does the "+1" come from? Is min() itself considered an illegal token? What token comes after max()? Where did you find this documented?

nyh replied:
I just noticed below you never call token_for_segment() for segment==0, so this + 1 doesn't even seem to make a difference...

psarna replied:
Indeed, + 1 doesn't make sense anymore with the special segment==0 code, it's a good idea and I'll drop it in v2. 

As for the other questions:
 * yes, we successfully deprecated all non-conforming partitioners, and token is now guaranteed to be 64 bits, which will allow keeping it per table and serializing/deserializing it without the knowledge of its partitioner - since it will always be representable as int64 underneath. Browse the mailing list with `partitioner` keyword and you'll see several related patches by PiotrJ and others.
 * +1 came from the fact that I glimpsed in another series on the mailing list that min() is a special case. I can't find it now, but it was one of PiotrJ's patches about using a correct int distribution when choosing a random token.

nyh commented on alternator/executor.cc:
```diff
@@ -2951,11 +2951,38 @@ static future<executor::request_return_type> do_query(schema_ptr schema,
 }
 
+static dht::token token_for_segment(int segment, int total_segments) {
+    assert(total_segments > 1 && segment < total_segments);
```
Maybe also assert segment >= 0 if this is also a requirement?
In any case I guess none of these asserts are critical, because you test these things right before calling this code.

psarna replied:
Yes, I can add one more, it makes sense. The only semi-critical assertion I originally put here was `total_segments > 0`, since I don't want to end the world by trying to divide by zero :) The other conditions were added later.

nyh commented on alternator/executor.cc:
```diff
@@ -2966,10 +2993,21 @@ future<executor::request_return_type> executor::scan(client_state& client_state,
+            return make_ready_future<request_return_type>(api_error("ValidationException",
+                    "Segment must be non-negative and less than TotalSegments"));
+        }
+        if (*total_segments < 0 || *total_segments > 1000000) {
```
I'm curious why we need this upper limit - does DynamoDB do it?

psarna replied:
Yes, for this one I can actually add a test case, since it was not guarded by boto3 validation. The range of allowed TotalSegments in DynamoDB is `1 - 1000000`. It's also in the docs, but since those are not always reliable, I also checked that manually.

psarna commented:
v2 changes:
 * added a small test case for incorrect parameters
 * removed the now unneeded `+ 1` from computing tokens for segments
 * added one more condition to the assertion

nyh commented:
Thanks. Committed to next. Closing manually because this isn't a merge commit so Github won't notice...
